Automattic/wp-calypso
# My Home: simplify My Home data layer by migrating to React Query

## Changes proposed in this Pull Request

In the spirit of trialing React Query more broadly (p4TIVU-9GP-p2) this PR removes the redux store for My Home (`/home/{{ site url }}`) and replaces it with a mixture of React Query and the existing user preferences store. My Home doesn't deal with a lot of client state and is basically a UI that reflects the API, so it felt like a good match for React Query.

* Move the open/close state for the Quick Links card to the user preferences store
* Move the layout fetching and skipping from Redux to ReactQuery
* Small tweak: don't require the `?dev=true` query param when using the `?view=VIEW_ID` option to force a particular view

### Move the open/close state for the Quick Links card to the user preferences store

That's the state for this toggle:
<img width="316" alt="toggle" src="https://user-images.githubusercontent.com/1500769/123233892-90d4a900-d52e-11eb-95de-45dc14bd6b08.png">

Previously the toggle state was being persisted locally in indexdb by virtue of being in the home redux store (`calypso/state/home/`). However we already have a place in redux for this sort of user state (`calypso/state/preferences/`).

One advantage is this preference will now be saved on the server rather than not being remembered across devices. Another advantage is preferences can be inspected using the dev tools:
<img width="692" alt="prefs" src="https://user-images.githubusercontent.com/1500769/123234910-7c44e080-d52f-11eb-8fb8-e6faaa83c577.png">

And the final advantage, is moving the preference allows us to delete the redux store for My Home.

This PR doesn't migrate the user's existing preference into the new store i.e. if the user had collapsed Quick Links, this PR will re-open it until they close it again manually. I initially had logic that migrated the preference across, but I didn't think it was worth the extra bytes and complexity for this particular preference. You can check it out here: #53766

### Move the layout fetching and skipping from Redux to ReactQuery

There's an introductory post on the Calypso P2 p4TIVU-9AF-p2, but the short version is that React Query uses hooks to manage server state. It deals with loading state, caching, exponential backoff after errors etc. Basically the stuff that `calypso/data-layer/` does in a Redux way, React Query does in a hooks way.

It's especially suited when a UI doesn't have any _real_ application state of its own, it's just showing the servers state. And I realised that this matched My Home.

The My Home layout data is queried using the `useHomeLayoutQuery()` hook, and views are skipped using the `useSkipCurrentViewMutation()` hook. These hooks are in `calypso/data/home/`, and you can find other React Query hooks in the `data/` folder.

Some implementation notes off the top of my head:
- React Query likes to automatically keep the data up to date, but the home layout endpoint returns a random result each time it's called, so I've used flags to make sure it will only update on user action
- I'm grabbing the `dev` and `view` url params in the hooks so that we don't have to plumb the props all through home components
- New layout requests always override the old layout, so the query key is set up to ensure there's only ever one layout per site

### No longer require the `?dev=true` query param when forcing a specific view

My Home shows the view that the server thinks is most appropriate for the user. For testing purposes it's useful to be able to force a particular view, which can be done with by adding query params to the browser url bar e.g. `/home/{{ site url }}?dev=true&view=VIEW_WEBINARS`

There didn't seem to be any reason why we should only allow the view to be forced when the `dev` flag is also set. So I've removed that requirement.

The dev flag isn't currently used for anything on the back end, but I'm currently thinking we should retain it because it'd be useful if we ever want to trial new views, or some other new behaviour, and allow a12s to test it without a sandbox.

## Testing instructions



Everything should operate as before, but here are some things to try:

* Refreshing My Home will display a view that the user is eligable to see
* Skipping a view skips the correct view (check the POST body to ensure the correct view is skipped) and results in a new view being shown
* Check in the network tab that there aren't more requests than expected (no double-requests for example)
* Check the `?dev=true` and `?view=VIEW_ID` query params are correctly sent
* `calypso_customer_home_card_impression` tracks event should still be sent
* Test that the Quick Links toggle state is remembered correctly
